### PR TITLE
[core] update to c++20

### PR DESCRIPTION
**Pull request by RedAuburn**
Seems to be all that's needed to upgrade. 

Using g++13 because it's the first version with support for std::format (and close to full support for c++20 generally afaik) https://stackoverflow.com/a/69156536

building android & desktop works

## Important limitations on different iOS and XCode versions for some C++17 and above features support

https://developer.apple.com/xcode/cpp/

### CODE REVIEW

**Comment by vng** [comment]
@pastk @Ferenc- Merge only after yours approve. Please, check at least the compilation. You have some fancy/old Linux distros, AFAIK.

File: docs/CPP_STYLE.md
Comment (vng) [inline docs/CPP_STYLE.md:7]: Let's give a respect to our fav language :) Write with capital C++20 here and below.

File: CMakeLists.txt
Comment (RedAuburn) [reply CMakeLists.txt:153]: yes 😅 i guess it could be added in that PR, but i thought it was just easier to add it now

File: docs/CPP_STYLE.md
Comment (RedAuburn) [reply docs/CPP_STYLE.md:7]: fixed :)

**Comment by RedAuburn** [comment]
fixed capitalisation in docs

**Comment by AndrewShkrob** [comment]
I've added the label `Coverage` to ensure that you remember to fix this runner as well :)

**Comment by biodranik** [comment]
After starting using any XCode 15 feature we break the option to build and run XCode 14 with iOS 12 simulator.

**Comment by pastk** [comment]
> 
> Using g++13 because it't the first version with support for std::format

GCC 13.1 was released last April, so its super new, e.g. its not the default in the current Ubuntu 22.04 LTS.

There are custom backports of GCC 12 to my very dated system, but not 13 (though maybe I didn't search thoroughly enough).

Anyway its very likely there will be building issues on many older systems.

My understanding is that its better to find some alternative to `std::format` support...

**Comment by RedAuburn** [comment]
hmm yeah it's sounding like an update to c++20 might be more trouble than it's worth

**Comment by biodranik** [comment]
Let's make a safe list of C++20 features that are already available on most of our operating systems, that we can benefit from, and save our development time. Enabling C++20 compilation doesn't mean that we are forced to use its features. We can choose only some convenient one/important one that help us to write cleaner code.

**Comment by biodranik** [comment]
@RedAuburn is snprintf enough for your task?

**Comment by AndrewShkrob** [comment]
> Let's make a safe list of C++20 features that are already available on most of our operating systems.

@biodranik please, do not overcomplicate. We have github runners to do this work for us:
 - The user creates a PR
 - Runner fails
 - Feature is not supported
 - Fix with c++17 features

If some of our regular contributors use outdated gcc/libs, then I suggest them to use our "official" supported gcc/libs from runners.
Or if that's not possible (which I doubt) we can add an additional configuration for runners to utilize their gcc/libs.

**Comment by Ferenc-** [comment]
> There are custom backports of GCC 12 to my very dated system, but not 13 (though maybe I didn't search thoroughly enough).

✔️  I have checked it is also available for your `18.04`:
````
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt update
sudo apt install gcc-13 g++-13
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-13 10 --slave /usr/bin/g++ g++ /usr/bin/g++-13
````

> Anyway its very likely there will be building issues on many older systems.

✔️  So far I haven't found any extra build issues, that would be tied to older systems.
We have been providing build instructions in the past in the documentation and we can continue to do so in the future.
Furthermore there is the option to build in a container / flatpak.

@RedAuburn would it be an option to keep this PR branch on top of the current upstream `master`? For example one fix pertaining to the desktop build [is already there](https://github.com/organicmaps/organicmaps/pull/7302). Some more will be needed for the tests.

**Comment by biodranik** [comment]
@AndrewShkrob did you see [the table for iOS](https://developer.apple.com/xcode/cpp/#c++20)? Just build error is not enough, there are also some runtime OS dependencies in some cases.

**Comment by pastk** [comment]
> ✔️ I have checked it is also available for your `18.04`:
> 

Many thanks, I'll give it a go later when I have time!

**Comment by RedAuburn** [comment]
(rebased)

**Comment by Ferenc-** [comment]
@RedAuburn please rebase again to  upstream `master`, there is one needed change there.
After that please consider `cherry-pick` -ing [this commit](https://github.com/Ferenc-/organicmaps/commit/1fb54705be63ccef38e34a7665cda86f89bf158a) from me, which is needed because of a runner issue (https://github.com/actions/runner-images/issues/8659). After all that the CI should be passing ✔️

**Comment by RedAuburn** [comment]
> @RedAuburn please rebase again to upstream `master`, there is one needed change there. After that please consider `cherry-pick` -ing [this commit](https://github.com/Ferenc-/organicmaps/commit/1fb54705be63ccef38e34a7665cda86f89bf158a) from me, which is needed because of a runner issue ([actions/runner-images#8659](https://github.com/actions/runner-images/issues/8659)). After all that the CI should be passing ✔️

rebased & set clang to 15 👍️ thank you for your work!

**Comment by RedAuburn** [comment]
> Then also in `.github/workflows/coverage-check.yaml`, there is still 14, that's why the last job is still failing.

ah, fixed 👍️

**Comment by AndrewShkrob** [comment]
iOS is still building with c++17:
https://github.com/organicmaps/organicmaps/blob/1bc3960c69b26b15e540641927416b025d020ef1/xcode/common.xcconfig#L67

**Comment by RedAuburn** [comment]
rebased & changed xcode to c++20 👍️

**Comment by Ferenc-** [comment]
@RedAuburn please rebase again.

**Comment by pastk** [comment]
> ✔️ I have checked it is also available for your `18.04`:

I've tested it works fine on my Ubuntu 18.04! Tested both release and debug desktop builds.

**Comment by pastk** [comment]
> ✔️ I have checked it is also available for your `18.04`:

I've installed gcc 13 onto my ubuntu 18.04 system, OM desktop (master branch) compiles fine with it (both release and debug builds).

Once this is rebased I'll also test how it works with C++20 features enabled in our code.

**Review by Ferenc- — APPROVED** [review]

**Comment by pastk** [comment]
> I've installed gcc 13 onto my ubuntu 18.04 system, OM desktop (master branch) compiles fine with it (both release and debug builds).
> 
> Once this is rebased I'll also test how it works with C++20 features enabled in our code.

Builds and runs well both on my desktop and android device!

**Comment by vng** [comment]
@kirylkaveryn 
Ok, contributors with XCode are needed. Everything works fine with this PR?

**Comment by kirylkaveryn** [comment]
> @kirylkaveryn Ok, contributors with XCode are needed. Everything works fine with this PR?

@vng 
Build and run successfully on the simulator, iPhone device and mac (iPad designed for mac).

But I've got a **crash** in this branch in iOS (on iPhone):
- open the search screen
- tap on categories and select some
- see the map with pois
- start dragging the map and after some time the app will crash

I don't investigate if this bug is related to this PR but the master doesn't crash...

I can create an issue if someone reproduces this crash. But I see it the first time and only here.

Screenshot:
image

And video:

https://github.com/organicmaps/organicmaps/assets/79797627/51ab9e88-f9be-4928-9969-f529eaab3e48

**Comment by biodranik** [comment]
There could be other surprises too. It may make sense to beta-test this branch first before merging (after fixing the crash, of course).

Maybe analyzing the whole project with the latest XCode on this branch can discover some issues that hint at the crash.

**Review by biodranik — CHANGES_REQUESTED** [review]
I'm putting it on hold until the crash is fixed and the PR is tested.

What is the minimum gcc version that will work with our code on this branch?

**Comment by vng** [comment]
I don't think this crash is connected to cpp-20. I saw this assert before.

**Comment by Ferenc-** [comment]
> What is the minimum gcc version that will work with our code on this branch?

@biodranik it is part of the changeset. See  [CMakeLists.txt](https://github.com/organicmaps/organicmaps/pull/7299/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR143):
````
# GCC 13 is required to support std::format in map/place_page_info.cpp                                                                                                                    
if (CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 13.0)                                                                                                 
  message(FATAL_ERROR "Minimum supported g++ version is 13.0 yours is ${CMAKE_CXX_COMPILER_VERSION}") 
````

Not counting this forward looking limitation, the current `master` branch easily works with at least gcc 11 (with `-std=c++20`) as well.

**Comment by biodranik** [comment]
Ubuntu 22 comes with 12, requiring 13 could be a deal breaker, no?

**Comment by Ferenc-** [comment]
> Ubuntu 22 comes with 12, requiring 13 could be a deal breaker, no?

IMHO an average user who doesn't use self compiled code, would use the flatpak, where this doesn't matter. Advanced users and developers [can follow the aformentioned steps](https://github.com/organicmaps/organicmaps/pull/7299#issuecomment-1925872784).

**Comment by Osyotr** [comment]
Is there something blocking this PR? I would like to enable C++20 for at least two reasons:
1) `std::chrono` additions to replace POSIX extensions.
1) `std::string_view` constructor from iterators to remove nasty hack in `TokenizeIteratorBase`.

**Comment by Ferenc-** [comment]
> Is there something blocking this PR? I would like to enable C++20 for at least two reasons:
> 
> 1. `std::chrono` additions to replace POSIX extensions.
> 2. `std::string_view` constructor from iterators to remove nasty hack in `TokenizeIteratorBase`.

I second that. The new Ubuntu LTS is out with gcc `13`, other distros have done so long ago.
IMHO at least we should at least start testing with `c++20` in the CI, because otherwise it will be a waste of effort to keep fixing up issues on `master`, that are not prevented by a CI.

**Comment by biodranik** [comment]
Nobody has yet built and tested mobile apps with C++20 on different devices. Any volunteers?

File: .github/workflows/linux-check.yaml
Comment (biodranik) [inline .github/workflows/linux-check.yaml:101]: Can compiler be updated independently from switching the code to C++20?

File: CMakeLists.txt
Comment (biodranik) [inline CMakeLists.txt:143]: Did anyone test std::format already on iOS 12 and Android 5 devices?

File: .github/workflows/linux-check.yaml
Comment (Osyotr) [reply .github/workflows/linux-check.yaml:101]: I'd say there's no reason to update GCC in this PR and Clang 15 is still too old anyways.

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 14

File: .github/workflows/linux-check.yaml
Comment (Ferenc-) [reply .github/workflows/linux-check.yaml:101]: `gcc 14` and `clang 18` are now in `master`. @RedAuburn please consider a rebase

Comment (biodranik) [reply .github/workflows/linux-check.yaml:101]: @Ferenc- this is an incorrect statement ) `gcc 14 and clang 18 are enabled on Github CI` is the correct one.

**Comment by Jean-BaptisteC** [comment]
Can your rebase your branch?
Maybe it's can be tested in the next beta

**Comment by biodranik** [comment]
Is there a list of C++20 features that are not supported on older Android 5 and iOS 12 devices and may lead to crashes or issues?

**Comment by Ferenc-** [comment]
> Is there a list of C++20 features that are not supported on older Android 5

I might be completely wrong here, but isn't the C++ runtime and all of its features carried onto Android within the app because of the `'-DANDROID_STL=c++_static'` in `android/app/build.gradle`? 
[Ref 1](https://cmake.org/cmake/help/latest/variable/CMAKE_ANDROID_STL_TYPE.html) [Ref 2](https://developer.android.com/ndk/guides/cpp-support#sr).

**Comment by biodranik** [comment]
There is also https://github.com/android/ndk/discussions/1981

And this note:
Note: For full details of the expected level of C++ library support for any given version, see the [C++14 Status](https://libcxx.llvm.org/Status/Cxx14.html), [C++17 Status](https://libcxx.llvm.org/Status/Cxx17.html), and [C++20 Status](https://libcxx.llvm.org/Status/Cxx20.html) pages. (C++20 was previously known as C++2a.) The level of C++ language support in the compiler is orthogonal; see [C++ Support in Clang](https://clang.llvm.org/cxx_status.html) instead.

C++ locale class, for example, was not supported from the beginning on Android. What else is missing?

Note that there's no static runtime on iOS, so that looks like the weakest spot now with iOS 12.

**Comment by biodranik** [comment]
Please rebase, and let's alpha-test it, need to make sure that iOS 12/13 works on the current code.

File: CMakeLists.txt
Comment (RedAuburn) [inline CMakeLists.txt:171]: this is correct, right?

**Comment by kirylkaveryn** [comment]
> Please rebase, and let's alpha-test it, need to make sure that iOS 12/13 and Android 5+ work on the current code.
> 
> Important iOS limitations: https://developer.apple.com/xcode/cpp/

I've tested this PR on ios 12.5 and didn't encounter any issues.

**Comment by RedAuburn** [comment]
fixing coverage check in https://github.com/organicmaps/organicmaps/pull/8308 (unrelated to this pr)

**Comment by RedAuburn** [comment]
working fine on Android 5.1 (Oneplus One)

File: CMakeLists.txt
Comment (biodranik) [reply CMakeLists.txt:171]: No, please check [this table](https://gcc.gnu.org/projects/cxx-status.html), and it would be great to also test on different GCC versions. Highly likely that gcc-10 or 11 would be ok. You can experiment locally in this PR by adding other gcc version matrix values in the .github workflow check.

File: docs/CPP_STYLE.md
Comment (Osyotr) [inline docs/CPP_STYLE.md:14]: Unfortunately, the filesystem remark is still relevant.
https://github.com/organicmaps/organicmaps/pull/8256#issuecomment-2134054267

Comment (RedAuburn) [reply docs/CPP_STYLE.md:14]: Damnit apple, will update

**Comment by RedAuburn** [comment]
gcc 10 works

**Comment by RedAuburn** [comment]
9.0 failed, going with minversion of 10

File: CMakeLists.txt
Comment (RedAuburn) [reply CMakeLists.txt]: checked, gcc-9 fails but gcc-10 works, have updated PR 👍️

**Comment by RedAuburn** [comment]
should be good to merge finally :)

File: docs/CPP_STYLE.md
Comment (biodranik) [reply docs/CPP_STYLE.md:14]: Elementary string conversions std::to_chars, std::from_chars is also not available.

Comment (RedAuburn) [reply docs/CPP_STYLE.md:14]: noted, thanks

**Review by biodranik — COMMENTED** [review]
Thanks, LGTM. Let's test it before the merge in alpha/beta.

CC @vng @rtsisyk

**Comment by RedAuburn** [comment]
added info that `std::to_chars` & `std::from_chars` aren't available to docs

**Comment by RedAuburn** [comment]
accidentally messed up dco signature, fixed

File: CMakeLists.txt
Comment (RedAuburn) [reply CMakeLists.txt]: removed

File: CMakeLists.txt
Comment (Ferenc-) [inline CMakeLists.txt:171]: How do we expect to maintain this minimum `10.0`, if we don't even test in the CI with `10.0`?
Next PR going in after this, can add something, which suddenly needs `14.0` and nobody even notices, because that still passes the CI.

Comment (Osyotr) [reply CMakeLists.txt:171]: Suggestion: remove this check altogether.

Comment (biodranik) [reply CMakeLists.txt:171]: This check will save us and other contributors time debugging build issues when older compilers are used.

We can switch from clang-18 to gcc-10 in the Linux check `Linux no unity build` job.

Comment (Osyotr) [reply CMakeLists.txt:171]: You can express your usage requirements in CMake using [target_compile_features](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html) together with things from [CMAKE_CXX_KNOWN_FEATURES](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html).
That way you don't need to check anything - CMake will do it for you.

Comment (biodranik) [reply CMakeLists.txt:171]: What exactly do you propose to add?

Comment (Osyotr) [reply CMakeLists.txt:171]: I don't propose to add anything, I just don't think this check solves any real problem.

Comment (biodranik) [reply CMakeLists.txt:171]: Waste of time is a _real_ problem, especially at scale.

Comment (Ferenc-) [reply CMakeLists.txt:171]: Unfurtunately `CMAKE_CXX_KNOWN_FEATURES`, doesn't have individual feature support for C++20, and even if it had I don't think it could really subsitute actually testing, and currently we only test something that we could call "upper bound" compilers.
I propose to either split or change the [linux-matrix](https://github.com/organicmaps/organicmaps/blob/master/.github/workflows/linux-check.yaml#L96), so it tests lower bound and upper bound compilers, not just the upper bound as it is currently.
In any case line 169  `# GCC 8.1 is required` should be changed in this PR.

File: CMakeLists.txt
Comment (biodranik) [inline CMakeLists.txt:169]:
```suggestion
# GCC 10.0 is required to support <charconv> header inclusion in base/string_utils.hpp
```

Is GCC version mentioned somewhere in the Linux build docs? Should we mention it explicitly there too?

**Review by biodranik — COMMENTED** [review]
Let's finish and merge this PR, looks like it has passed alpha testing. There is one report about strange crashes on Android, it is not clear yet if it is related to this or other changes.

File: CMakeLists.txt
Comment (Ferenc-) [reply CMakeLists.txt:169]: > Should we mention it explicitly there too?

Yes, but IMHO that can also go into a separate PR.

**Review by biodranik — APPROVED** [review]
Ok, let's upgrade our _core_, and learn new patterns! Thanks to everyone involved, let C++20 be in OM!

**Pull request merged**
